File: 백준/2805.py
# 4 7
# 20 15 10 17
n, m = map(int, input().split())
trees = list(map(int, input().split()))

left = 0
right = max(trees)
answer = 0  # 높이의 최댓값

while left <= right:
    mid = (left + right) // 2
    cnt = 0
    for tree in trees:
        if tree - mid > 0:  # tree가 더 작으면 무시
            cnt += tree - mid
    if cnt >= m:
        answer = max(answer, mid)
        left = mid + 1
    else:
        right = mid - 1
print(answer)

File: milla/2805.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
int main(void){
    // 4 7 
    // 20 15 10 17
    long long N, M;
    cin >> N >> M;
    long long trees[N];
    for (int i=0;i<N;i++){
        cin >> trees[i];
    }
    long long answer=0;
    long long left = 1;
    long long right = *max_element(trees, trees+N);
    while(left <=right){
        long long point=0;
        long long mid = (left+right)/2;
        //cout << "mid" << mid << endl;
        for (long long i=0;i<N;i++){
            if (trees[i]>mid){
                point+=trees[i]-mid;
            }
        }
        //cout << "point" << point << endl;
        if(point>=M){
            left = mid+1;
            answer=mid;
        }
        else{
            right=mid-1;
        }
    }
    cout << answer;
}